protocols/primary-backup/src/main/java/io/atomix/protocols/backup/roles/AsynchronousReplicator.java

```diff
@@ -110,7 +118,7 @@ private void backup() {
           context.currentIndex(),
           batch);
       log.trace("Sending {} to {}", request, nodeId);
-      context.protocol().backup(nodeId, request);
+      context.protocol().backup(nodeId, request).join();
```

Comment:
Instead of join here can we change the method signature to return completable future and return the cf from backup then propagate that up the call stack?

Reply:
at least for add it's quite easy..

```java
  @Override
  public CompletableFuture<Void> replicate(BackupOperation operation) {
    List<CompletableFuture<Void>> futures = new ArrayList<>(context.backups().size());
    for (NodeId backup : context.backups()) {
      futures.add(queues.computeIfAbsent(backup, BackupQueue::new).add(operation));
    }
    context.setCommitIndex(operation.index());
    return Futures.allOf(futures).thenApply(v -> null);
  }
```

Reply:
this may make ```replicate``` same as ```SynchronousReplicator```